File: 001-sorting/counting-sort/counting-sort.js
function getMaxElement(a) {
  var max = a[0];
  for (var i = 1; i < a.length; i++) {
    if (a[i] > max) {
      max = a[i];
    }
  }
  return max;
}

function countingSort(a, k) {
  var c = [];
  var b = [];
  for (var i = 0; i <= k; i++) {
    c[i] = 0;
  }

  for (var i = 0; i < a.length; i++) {
    c[a[i]] = c[a[i]] + 1;
  }

  for (var i = 1; i < c.length; i++) {
    c[i] = c[i - 1] + c[i];
  }

  for (var i = a.length - 1; i >= 0; i--) {
    b[c[a[i]] - 1] = a[i];
    c[a[i]] = c[a[i]] - 1;
  }
  return b;
}

var a = [10, 23, 1, 5, 23, 10, 67, 8, 9, 11, 45, 14];
console.log("Before sorting", a);
var k = getMaxElement(a);
a = countingSort(a, k);
console.log("After sorting", a);

File: 001-sorting/counting-sort/counting-sort.go
package main

import "fmt"

func main() {
	a := []uint64{10, 9, 3, 8, 4, 0, 23, 18, 1, 0, 24}
	fmt.Println("Before sorting", a)
	k := getMaxElement(a)
	b := countingSort(a, k)
	fmt.Println("After sorting", b)
}

func getMaxElement(a []uint64) uint64 {
	max := a[0]
	for i := 1; i < len(a); i++ {
		if a[i] > max {
			max = a[i]
		}
	}
	return max
}

func countingSort(a []uint64, k uint64) []uint64 {
	c := make([]uint64, k+1, k+1)
	b := make([]uint64, len(a), len(a))
	for _, v := range a {
		c[v] = c[v] + 1
	}

	for i := 1; i < len(c); i++ {
		c[i] = c[i-1] + c[i]
	}

	for i := len(a) - 1; i >= 0; i-- {
		b[c[a[i]]-1] = a[i]
		c[a[i]] = c[a[i]] - 1
	}
	return b

}
